### PR TITLE
Fix missing components in pattern patterns and default search value for string pattern

### PR DESCRIPTION
**Pull request by juliandescottes**
Attempts to fix a few issues:
- when no value is provided for protocol, pathname or search in a URLPatternPattern, we should not attempt to match against it
- with a URLPatternString, the search component should default to empty string, not null

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -4684,26 +4702,47 @@ To <dfn>parse URL pattern</dfn>, given |pattern|:
...
+          [=default port=] is not null, and |url|'s [=url/port=] is null or is
+          equal to [=url/scheme=]'s [=default port=]:
 
-     1. Otherwise:
+         1. The empty string.
 
-       1. null.
+       1. Otherwise, if |url|'s [=url/port=] is not null:
+
+         1. [=Serialize an integer=] with |url|'s [=url/port=].
```
Comment (thiagowfx) [inline index.bs:4719]: Do you also need to mention here the integer range for a valid port number?

Comment (juliandescottes) [reply index.bs:4719]: At the moment we don't have to: this will already raise an error when trying to parse the URL, near the end of the parsing step. 

Now if we stop using the basic URL parser here (and it might make sense based on https://github.com/w3c/webdriver-bidi/issues/507#issuecomment-1688462008), adding this validation would make sense.

**Comment by juliandescottes** [comment]
cc @jgraham for review

**Comment by juliandescottes** [comment]
Small review ping for @jgraham

File: index.bs [hunk 2]
```diff
@@ -4684,26 +4702,47 @@ To <dfn>parse URL pattern</dfn>, given |pattern|:
...
      <dt>search
-     <dd>|url|'s [=url/query=].
+     <dd>
+     1. If |has search| is false:
+
+       1. null.
+
+     1. Otherwise:
+
+       1. The empty string if |url|'s [=url/path=] is null, or |url|'s [=url/query=] otherwise.
```
Comment (jgraham) [inline index.bs:4745]: Why do we set the empty string if the path is null?

Comment (juliandescottes) [reply index.bs:4745]: The idea is to match the [URLPattern's spec match algorithm](https://wicg.github.io/urlpattern/#match), which defaults to an empty string if the url's search is null. 

Basically this algorithm will never differentiate between empty string search or no search, so I always turn null into empty string, both here, and when reading the components of the URL to match.

Comment (juliandescottes) [reply index.bs]: I missed that this was `path` instead of `query` should be fixed now!

**Review by jgraham — APPROVED** [review]

**Comment by juliandescottes** [comment]
@thiagowfx maybe you want to take another look at this for a final review?

File: index.bs [hunk 3]
```diff
@@ -4748,8 +4787,13 @@ To <dfn>match URL pattern</dfn> given |url pattern| and |url string|:
 1. If |url pattern|'s pathname is not null and is not equal to the result of
    running the [=URL path serializer=] with |url|, return false.
 
-1. If |url pattern|'s search is not null and is not equal to |url|'s
-   [=url/query=], return false.
+1. If |url pattern|'s search is not null:
+
+   1. Let |url query| be |url|'s [=url/query=]
```
Comment (thiagowfx) [inline index.bs:4792]:
```suggestion
   1. Let |url query| be |url|'s [=url/query=].
```

**Review by thiagowfx — APPROVED** [review]
Are you planning to add WPT tests for these edge cases?

**Comment by juliandescottes** [comment]
Thanks for the review! 

> Are you planning to add WPT tests for these edge cases?

I tried to cover it already url_patterns_tentative.py (that's why we have the "tentative" suffix actually). Once this PR lands I will remove the suffix. That will happen downstream at the same time as I update our implementation to match the new spec: https://bugzilla.mozilla.org/show_bug.cgi?id=1849743
